src/world.rs

```diff
@@ -1,100 +1,67 @@
-}
-
-
-impl rand::Rand for CellState {
-    #[inline]
-    fn rand<R:rand::Rng>(_: &mut R) -> CellState {
-        match rand::random() {
-            true => CellState::Alive,
-            false => CellState::Dead
-        }
-    }
-}
-
-
 #[derive(Debug, Clone)]
 pub struct World {
     width: isize,
     height: isize,
-    cells: HashMap<Point, CellState>
+    generation: HashSet<Point>,
+    neighbours: HashMap<Point, usize>,
+    age: usize,
 }
 
-
 impl World {
     pub fn new(width: isize, height: isize)  -> World {
-        let mut world = World{width: width, height: height, cells: HashMap::new()};
-        for y in 0..height { for x in 0..width {
-            world.cells.insert((x, y), CellState::Dead);
-        }}
-        world
+        World { width: width, height: height, age: 0, generation: HashSet::new(),
+                neighbours: HashMap::with_capacity((width * height * 8) as usize) }
     }
 
-    pub fn evolve(&self) -> World {
-        let mut new_world = self.clone();
-        for y in 0..self.height { for x in 0..self.width {
-            new_world.set_state(x, y, self.decide_fate(x, y))
-        }}
-        new_world
-
+    pub fn seed(mut self) -> World {
+        self.generation = (0..self.width).cartesian_product(0..self.height)
+                                         .filter(|_| random())
+                                         .collect::<HashSet<Point>>();
+        self.calculate_neighbours();
+        self
     }
 
-    pub fn randomize(&mut self) {
-        for y in 0..self.height { for x in 0..self.width {
-            self.set_state(x, y, rand::random());
-        }}
-
-    }
-
-    fn get_sibling_coords(&self, x: isize, y: isize) -> Vec<Point> {
-        let mut coords = vec![];
-        for _x in [x + 1, x - 1, x].iter() {for _y in [y + 1, y - 1, y].iter() {
-            if (*_x != x || *_y != y) && 0 <= *_x && *_x < self.width && 0 <= *_y && *_y < self.height {
-                coords.push((*_x, *_y));
-            }
-        }}
-        coords
+    pub fn evolve(&mut self) -> &World {
+        self.generation = self.neighbours.iter()
+                              .filter(|&(n, c)| self.decide_fate(n, *c))
+                              .map(|(n, _)| *n)
+                              .collect::<HashSet<Point>>();
+        self.calculate_neighbours();
+        self.age += 1;
+        self
     }
 
-    fn decide_fate(&self, x: isize, y: isize) -> CellState {
-        let mut num_alive_siblings = 0;
-        for (_x, _y) in self.get_sibling_coords(x, y) {
-            if self.is_alive(_x, _y) {
-                num_alive_siblings += 1
+    fn calculate_neighbours(&mut self) {
+        self.neighbours.clear();
+        for &(x, y) in self.generation.iter() {
+            let neighbours = (-1..2).cartesian_product(-1..2)
+                                    .filter(|&(dx, dy)| dx != 0 || dy != 0)
+                                    .map(|(dx, dy)| (dx + x, dy + y));
+            for n in neighbours {
```

Comment:
I wanted to write this stuff using fold but unfortunately could not ((( ```.fold(self.neighbours, |mut acc, p| {  *self.neighbours.entry(n).or_insert(0) += 1; acc }```. Maybe you can find more good looking solution instead of this ugly for loop?